### PR TITLE
Fix a regression with scalar indexing due to #1800

### LINKED ISSUE #1874: Regression in v2.18.0 when setting an item in an object array
### Zarr version

v2.18.0

### Numcodecs version

0.12.1

### Python Version

3.12

### Operating System

Windows

### Installation

via `poetry install`

### Description

There seems to be a regression in v2.18.0 (compared to v2.17.2) concerning setting items in an object array.

### Steps to reproduce

Setup:

```py
import zarr
from numcodecs import MsgPack
import numpy as np

root = zarr.group()
arr = root.create_dataset(
    name="my_dataset",
    shape=0,
    dtype=object,
    object_codec=MsgPack(),
)
new_items = [
    ["A", 1],
    ["B", 2, "hello"],
]
arr_add = np.empty(len(new_items), dtype=object)
arr_add[:] = new_items
arr.append(arr_add)
arr[:] # --> array([list(['A', 1]), list(['B', 2, 'hello'])], dtype=object)
```
Problem:

```py
arr[0] = ["C", 3] # raises AttributeError: 'list' object has no attribute 'astype' (from Array._process_for_setitem)
```

Using Zarr v2.17.2, no such exception is raised, and the first item is successfully modified.

### Additional output

_No response_

### LINKED ISSUE #1883: Regression in 2.18: Index to size-1 array fails
### Zarr version

v2.18.0

### Numcodecs version

v0.12.1

### Python Version

3.11

### Operating System

Reproduced on mac and linux

### Installation

micromamba / conda-forge

### Description

Indexing to unit-length arrays fails since zarr 2.18. Indexing other array lengths still works. See reproducer script and traceback below. In zarr 2.17, this still worked. I don't understand the problem exactly, but based on the 2.18 changelog, I would guess that this is somehow related to https://github.com/zarr-developers/zarr-python/pull/1800

### Steps to reproduce
```python
import zarr
import numcodecs

assert zarr.__version__ == "2.18.0"
assert numcodecs.__version__ == "0.12.1"

x = zarr.array([1, 2])
y = zarr.array([1])

# OK
print(x[:])
print(y[:])

# OK
print(x[0])

# Not OK
print(y[0])
```

### Additional output

```
Traceback (most recent call last):
  File "/Users/x/Library/Application Support/JetBrains/PyCharm2023.3/scratches/scratch_42.py", line 14, in <module>
    print(x[0])
          ~^^^
  File "/Users/x/micromamba/envs/envname/lib/python3.11/site-packages/zarr/core.py", line 800, in __getitem__
    result = self.get_basic_selection(pure_selection, fields=fields)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/x/micromamba/envs/envname/lib/python3.11/site-packages/zarr/core.py", line 926, in get_basic_selection
    return self._get_basic_selection_nd(selection=selection, out=out, fields=fields)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/x/micromamba/envs/envname/lib/python3.11/site-packages/zarr/core.py", line 968, in _get_basic_selection_nd
    return self._get_selection(indexer=indexer, out=out, fields=fields)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/x/micromamba/envs/envname/lib/python3.11/site-packages/zarr/core.py", line 1343, in _get_selection
    self._chunk_getitems(
  File "/Users/x/micromamba/envs/envname/lib/python3.11/site-packages/zarr/core.py", line 2181, in _chunk_getitems
    self._process_chunk(
  File "/Users/x/micromamba/envs/envname/lib/python3.11/site-packages/zarr/core.py", line 2049, in _process_chunk
    self._compressor.decode(cdata, dest)
  File "numcodecs/blosc.pyx", line 564, in numcodecs.blosc.Blosc.decode
  File "numcodecs/blosc.pyx", line 365, in numcodecs.blosc.decompress
  File "numcodecs/compat_ext.pyx", line 16, in numcodecs.compat_ext.Buffer.__cinit__
ValueError: buffer source array is read-only
```

### PR DESCRIPTION
**Pull request by dcherian**
Fixes a regression in indexing out scalars from size-1 chunks.

TODO:
* [x] Closes #1874, Closes #1883
* [x] Add unit tests and/or doctests in docstrings
* [x] Changes documented in docs/release.rst
* [x] GitHub Actions have all passed
* [x] Test coverage is 100% (Codecov passes)

### CODE REVIEW

**Review by rabernat — APPROVED** [review]
Thanks Deepak!

File: zarr/indexing.py [hunk 1]
```diff
@@ -52,6 +52,8 @@ def is_scalar(value, dtype):
         return True
     if isinstance(value, tuple) and dtype.names and len(value) == len(dtype.names):
         return True
+    if dtype.kind == "O" and not isinstance(value, np.ndarray):
```
Comment (dcherian) [inline zarr/indexing.py:55]: This is the core issue in #1874. We need to treat `[1, 2]` as a scalar for object arrays.

Is this the right condition?

cc @jakirkham @aplowman

**Review by jhamman — APPROVED** [review]

**Comment by jhamman** [comment]
@zarr-developers/python-core-devs - I'm planning to merge this and make a 2.18.1 release later today.

**Review by rabernat — APPROVED** [review]
Thanks Deepak!

**Pull request merged**